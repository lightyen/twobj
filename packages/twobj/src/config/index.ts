export { resolveConfig } from "./resolveConfig"
import config from "./defaultConfig"
import colors from "./defaultColors"

const defaultConfig = config
const defaultColors = colors

export { defaultConfig, defaultColors }
