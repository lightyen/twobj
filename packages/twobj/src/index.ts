export { resolveConfig } from "./config/resolveConfig"
export { createContext } from "./core"
export type { CSSProperties, CSSValue } from "./types"
export type { ValueType } from "./values"
import type { CSSProperties } from "./types"

// Fake
export const globalStyles: Record<string, CSSProperties> = {}
export function tw(arr: TemplateStringsArray): CSSProperties {
	return {}
}
export function theme(arr: TemplateStringsArray): unknown {
	return undefined
}
