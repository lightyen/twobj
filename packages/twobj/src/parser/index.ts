export * from "./angle"
export * from "./css"
export * from "./length"
export * from "./nodes"
export * from "./number"
export * from "./parser"
export * from "./percentage"
export * from "./shadow"
export * from "./theme"
export * from "./util"
