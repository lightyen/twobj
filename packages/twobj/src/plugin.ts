/* eslint-disable @typescript-eslint/no-explicit-any */
import type { ConfigJS, CorePlugin, UnnamedPlugin, UserPluginFunctionWithOption, UserPluginObject } from "./types"

interface PluginFunction<Options> {
	(options: Options): UnnamedPlugin
}
interface ConfigFunction<Options> {
	(options: Options): ConfigJS
}

export interface CreatePlugin {
	(handler: UnnamedPlugin): CorePlugin
	(pluginName: string, handler: UnnamedPlugin): CorePlugin
	(handler: UnnamedPlugin, config: ConfigJS): CorePlugin
	withOptions: CreatePluginWithOptions
}

interface CreatePluginWithOptions {
	<Options = unknown>(
		pluginFunction: PluginFunction<Options>,
		configFunction?: ConfigFunction<Options>,
	): UserPluginFunctionWithOption<Options>
	<Options = unknown>(
		pluginName: string,
		pluginFunction: PluginFunction<Options>,
		configFunction?: ConfigFunction<Options>,
	): UserPluginFunctionWithOption<Options>
}

/** Create a tailwind plugin. */
export const plugin: CreatePlugin = (first: string | UnnamedPlugin, second?: unknown, third?: unknown) => {
	// prevent plugin name to be minified
	if (typeof first === "string") {
		Object.defineProperty(second, "name", { value: first, writable: false })
		return second as CorePlugin
	}
	return first
}

plugin.withOptions = function <Options = unknown>(
	first: string | ((options: Options) => UnnamedPlugin),
	second: unknown,
	third?: unknown,
): UserPluginFunctionWithOption<Options> {
	const pluginName = typeof first === "string" ? first : ""

	let pluginFunction: PluginFunction<Options>
	let configFunction: ConfigFunction<Options>
	if (typeof first === "string") {
		pluginFunction = second as PluginFunction<Options>
		configFunction = (third ?? (() => ({}))) as ConfigFunction<Options>
	} else {
		pluginFunction = first
		configFunction = (second ?? (() => ({}))) as ConfigFunction<Options>
	}

	function optionsFunction(options: Options): UserPluginObject {
		return {
			name: pluginName,
			handler: pluginFunction(options),
			config: configFunction(options),
		}
	}

	optionsFunction.__pluginFunction = pluginFunction
	optionsFunction.__configFunction = configFunction
	optionsFunction.__isOptionsFunction = true
	return optionsFunction
}
