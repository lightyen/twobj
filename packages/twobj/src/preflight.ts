import { CSSProperties } from "./types"

export const preflight: CSSProperties = {
	"*,::before,::after": {
		boxSizing: "border-box",
		borderWidth: "0",
		borderStyle: "solid",
		borderColor: "theme('borderColor.DEFAULT', currentColor)",
	},
	"::before,::after": { "--tw-content": "''" },
	html: {
		lineHeight: "1.5",
		WebkitTextSizeAdjust: "100%",
		MozTabSize: "4",
		tabSize: "4",
		fontFamily: `theme('fontFamily.sans', ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji")`,
	},
	body: { margin: "0", lineHeight: "inherit" },
	hr: { height: "0", color: "inherit", borderTopWidth: "1px" },
	"abbr:where([title])": { textDecoration: "underline dotted" },
	"h1,h2,h3,h4,h5,h6": { fontSize: "inherit", fontWeight: "inherit" },
	a: { color: "inherit", textDecoration: "inherit" },
	"b,strong": { fontWeight: "bolder" },
	"code,kbd,samp,pre": {
		fontFamily: `theme('fontFamily.mono', ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace)`,
		fontSize: "1em",
	},
	small: { fontSize: "80%" },
	"sub,sup": {
		fontSize: "75%",
		lineHeight: "0",
		position: "relative",
		verticalAlign: "baseline",
	},
	sub: { bottom: "-0.25em" },
	sup: { top: "-0.5em" },
	table: {
		textIndent: "0",
		borderColor: "inherit",
		borderCollapse: "collapse",
	},
	"button,input,optgroup,select,textarea": {
		fontFamily: "inherit",
		fontSize: "100%",
		fontWeight: "inherit",
		lineHeight: "inherit",
		color: "inherit",
		margin: "0",
		padding: "0",
	},
	"button,select": { textTransform: "none" },
	"button,[type='button'],[type='reset'],[type='submit']": {
		WebkitAppearance: "button",
		backgroundColor: "transparent",
		backgroundImage: "none",
	},
	":-moz-focusring": { outline: "auto" },
	":-moz-ui-invalid": { boxShadow: "none" },
	progress: { verticalAlign: "baseline" },
	"::-webkit-inner-spin-button,::-webkit-outer-spin-button": { height: "auto" },
	"[type='search']": { WebkitAppearance: "textfield", outlineOffset: "-2px" },
	"::-webkit-search-decoration": { WebkitAppearance: "none" },
	"::-webkit-file-upload-button": { WebkitAppearance: "button", font: "inherit" },
	summary: { display: "list-item" },
	"blockquote,dl,dd,h1,h2,h3,h4,h5,h6,hr,figure,p,pre": { margin: "0" },
	fieldset: { margin: "0", padding: "0" },
	legend: { padding: "0" },
	"ol,ul,menu": { listStyle: "none", margin: "0", padding: "0" },
	textarea: { resize: "vertical" },
	"input::placeholder,textarea::placeholder": { opacity: "1", color: "theme('colors.gray.400', #9ca3af)" },
	'button,[role="button"]': { cursor: "pointer" },
	":disabled": { cursor: "default" },
	"img,svg,video,canvas,audio,iframe,embed,object": { display: "block", verticalAlign: "middle" },
	"img,video": { maxWidth: "100%", height: "auto" },
	"[hidden]": { display: "none" },
}
