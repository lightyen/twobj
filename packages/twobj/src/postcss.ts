import cssesc from "cssesc"
import * as parsel from "parsel-js"
import { escapeRegexp, splitAtTopLevelOnly } from "./parser"

// Postcss source: https://github.com/postcss/postcss-selector-parser/blob/master/src/util/unesc.js

export function escapeCss(value: string): string {
	return cssesc(value, { isIdentifier: true })
}

function gobbleHex(str: string): [string, number] | undefined {
	const lower = str.toLowerCase()
	let hex = ""
	let spaceTerminated = false
	for (let i = 0; i < 6 && lower[i] !== undefined; i++) {
		const code = lower.charCodeAt(i)
		// check to see if we are dealing with a valid hex char [a-f|0-9]
		const valid = (code >= 97 && code <= 102) || (code >= 48 && code <= 57)
		// https://drafts.csswg.org/css-syntax/#consume-escaped-code-point
		spaceTerminated = code === 32
		if (!valid) {
			break
		}
		hex += lower[i]
	}

	if (hex.length === 0) {
		return undefined
	}
	const codePoint = parseInt(hex, 16)

	const isSurrogate = codePoint >= 0xd800 && codePoint <= 0xdfff
	// Add special case for
	// "If this number is zero, or is for a surrogate, or is greater than the maximum allowed code point"
	// https://drafts.csswg.org/css-syntax/#maximum-allowed-code-point
	if (isSurrogate || codePoint === 0x0000 || codePoint > 0x10ffff) {
		return ["\uFFFD", hex.length + (spaceTerminated ? 1 : 0)]
	}

	return [String.fromCodePoint(codePoint), hex.length + (spaceTerminated ? 1 : 0)]
}

const CONTAINS_ESCAPE = /\\/
export function unescapeCss(str: string) {
	const needToProcess = CONTAINS_ESCAPE.test(str)
	if (!needToProcess) {
		return str
	}
	let ret = ""

	for (let i = 0; i < str.length; i++) {
		if (str[i] === "\\") {
			const gobbled = gobbleHex(str.slice(i + 1, i + 7))
			if (gobbled !== undefined) {
				ret += gobbled[0]
				i += gobbled[1]
				continue
			}

			// Retain a pair of \\ if double escaped `\\\\`
			// https://github.com/postcss/postcss-selector-parser/commit/268c9a7656fb53f543dc620aa5b73a30ec3ff20e
			if (str[i + 1] === "\\") {
				ret += "\\"
				i++
				continue
			}

			// if \\ is at the end of the string retain it
			// https://github.com/postcss/postcss-selector-parser/commit/01a6b346e3612ce1ab20219acc26abdc259ccefb
			if (str.length === i + 1) {
				ret += str[i]
			}
			continue
		}

		ret += str[i]
	}

	return ret
}

export function findClasses(selector: string): Map<string, string> {
	interface OtherNode {
		type: string
		list?: Node[]
		subtree?: Node
		left?: Node
		right?: Node
	}

	interface ClassNode {
		type: "class"
		content: string
		name: string
		pos: [number, number]
	}

	type Node = ClassNode & OtherNode

	const classes = new Map<string, string>()

	splitAtTopLevelOnly(selector).forEach(s => walk(parsel.parse(s.value)))

	return classes

	function isExist(node: parsel.AST | null | undefined): node is parsel.AST {
		return typeof node === "object" && node != null
	}

	function isTokens(node: parsel.AST): node is parsel.Tokens {
		switch (node.type) {
			case "complex":
			case "compound":
			case "list":
				return false
			default:
				return true
		}
	}

	function isList(node: parsel.AST): node is parsel.List {
		return node.type === "list" || Array.isArray(node["list"])
	}

	function isComplex(node: parsel.AST): node is parsel.Complex {
		return node.type === "complex" || node["left"] != null || node["right"] != null
	}

	function walk(node: parsel.AST | null | undefined) {
		if (!isExist(node)) {
			return
		}

		if (callback(node) === false) {
			return
		}

		if (isList(node)) {
			node.list.forEach(walk)
			return
		}

		if (isTokens(node)) {
			walk(node.subtree)
			return
		}

		if (isComplex(node) && isExist(node.left)) {
			walk(node.left)
		}

		if (isComplex(node) && isExist(node.right)) {
			walk(node.right)
		}
	}

	function callback(node: parsel.AST): boolean | void {
		if (node.type === "class") {
			if (!classes.has(node.name)) {
				const replaced = selector.replace(new RegExp(`[.]${escapeRegexp(node.name)}(?!-)\\b`, "g"), "&")
				classes.set(unescapeCss(node.name), replaced)
			}
			return false
		}
	}
}
