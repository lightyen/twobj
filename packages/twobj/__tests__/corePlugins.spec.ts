import { createContext, resolveConfig } from "../src"

test("preflight", async () => {
	expect(
		createContext(
			resolveConfig({
				corePlugins: {
					preflight: false,
				},
			}),
		).globalStyles,
	).toEqual({
		"*, ::before, ::after": {
			"--tw-backdrop-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-opacity": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-border-spacing-x": "0",
			"--tw-border-spacing-y": "0",
			"--tw-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-drop-shadow": "var(--tw-empty,/**/ /**/)",
			"--tw-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-figure": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-fraction": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-spacing": "var(--tw-empty,/**/ /**/)",
			"--tw-ordinal": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-x": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-y": "var(--tw-empty,/**/ /**/)",
			"--tw-pinch-zoom": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-color": "rgb(59 130 246 / 0.5)",
			"--tw-ring-inset": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-offset-color": "#fff",
			"--tw-ring-offset-shadow": "0 0 #0000",
			"--tw-ring-offset-width": "0px",
			"--tw-ring-shadow": "0 0 #0000",
			"--tw-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-scroll-snap-strictness": "proximity",
			"--tw-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-slashed-zero": "var(--tw-empty,/**/ /**/)",
			"--tw-transfrom-translate-default": "translate(var(--tw-translate-x, 0), var(--tw-translate-y, 0))",
		},
		"::backdrop": {
			"--tw-backdrop-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-opacity": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-border-spacing-x": "0",
			"--tw-border-spacing-y": "0",
			"--tw-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-drop-shadow": "var(--tw-empty,/**/ /**/)",
			"--tw-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-figure": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-fraction": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-spacing": "var(--tw-empty,/**/ /**/)",
			"--tw-ordinal": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-x": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-y": "var(--tw-empty,/**/ /**/)",
			"--tw-pinch-zoom": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-color": "rgb(59 130 246 / 0.5)",
			"--tw-ring-inset": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-offset-color": "#fff",
			"--tw-ring-offset-shadow": "0 0 #0000",
			"--tw-ring-offset-width": "0px",
			"--tw-ring-shadow": "0 0 #0000",
			"--tw-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-scroll-snap-strictness": "proximity",
			"--tw-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-slashed-zero": "var(--tw-empty,/**/ /**/)",
			"--tw-transfrom-translate-default": "translate(var(--tw-translate-x, 0), var(--tw-translate-y, 0))",
		},
	})

	expect(createContext(resolveConfig({})).globalStyles).toMatchSnapshot()

	expect(createContext(resolveConfig({})).globalStyles).toMatchObject({
		"*, ::before, ::after": {
			"--tw-backdrop-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-opacity": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-border-spacing-x": "0",
			"--tw-border-spacing-y": "0",
			"--tw-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-drop-shadow": "var(--tw-empty,/**/ /**/)",
			"--tw-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-figure": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-fraction": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-spacing": "var(--tw-empty,/**/ /**/)",
			"--tw-ordinal": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-x": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-y": "var(--tw-empty,/**/ /**/)",
			"--tw-pinch-zoom": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-color": "rgb(59 130 246 / 0.5)",
			"--tw-ring-inset": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-offset-color": "#fff",
			"--tw-ring-offset-shadow": "0 0 #0000",
			"--tw-ring-offset-width": "0px",
			"--tw-ring-shadow": "0 0 #0000",
			"--tw-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-scroll-snap-strictness": "proximity",
			"--tw-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-slashed-zero": "var(--tw-empty,/**/ /**/)",
			"--tw-transfrom-translate-default": "translate(var(--tw-translate-x, 0), var(--tw-translate-y, 0))",
		},
		"::backdrop": {
			"--tw-backdrop-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-opacity": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-backdrop-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-blur": "var(--tw-empty,/**/ /**/)",
			"--tw-border-spacing-x": "0",
			"--tw-border-spacing-y": "0",
			"--tw-brightness": "var(--tw-empty,/**/ /**/)",
			"--tw-contrast": "var(--tw-empty,/**/ /**/)",
			"--tw-drop-shadow": "var(--tw-empty,/**/ /**/)",
			"--tw-grayscale": "var(--tw-empty,/**/ /**/)",
			"--tw-hue-rotate": "var(--tw-empty,/**/ /**/)",
			"--tw-invert": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-figure": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-fraction": "var(--tw-empty,/**/ /**/)",
			"--tw-numeric-spacing": "var(--tw-empty,/**/ /**/)",
			"--tw-ordinal": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-x": "var(--tw-empty,/**/ /**/)",
			"--tw-pan-y": "var(--tw-empty,/**/ /**/)",
			"--tw-pinch-zoom": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-color": "rgb(59 130 246 / 0.5)",
			"--tw-ring-inset": "var(--tw-empty,/**/ /**/)",
			"--tw-ring-offset-color": "#fff",
			"--tw-ring-offset-shadow": "0 0 #0000",
			"--tw-ring-offset-width": "0px",
			"--tw-ring-shadow": "0 0 #0000",
			"--tw-saturate": "var(--tw-empty,/**/ /**/)",
			"--tw-scroll-snap-strictness": "proximity",
			"--tw-sepia": "var(--tw-empty,/**/ /**/)",
			"--tw-slashed-zero": "var(--tw-empty,/**/ /**/)",
			"--tw-transfrom-translate-default": "translate(var(--tw-translate-x, 0), var(--tw-translate-y, 0))",
		},
	})
})

test("disable corePlugins", async () => {
	const ctx = createContext(
		resolveConfig({
			corePlugins: {
				backgroundColor: false,
			},
		}),
	)
	expect(ctx.css("bg-black")).toEqual({})
	expect(ctx.css("bg-auto")).toEqual({ backgroundSize: "auto" })
})

test("disable all corePlugins", async () => {
	let ctx = createContext(
		resolveConfig({
			corePlugins: [],
		}),
	)
	expect(ctx.css("bg-black")).toEqual({})
	expect(ctx.css("bg-auto")).toEqual({})

	ctx = createContext(
		resolveConfig({
			corePlugins: false,
		}),
	)
	expect(ctx.css("bg-black")).toEqual({})
	expect(ctx.css("bg-auto")).toEqual({})
})

test("addVariant", async () => {
	const ctx = createContext(
		resolveConfig({
			plugins: [
				({ addVariant }) => {
					addVariant("prose-headings", ":where(&) :is(h1, h2, h3, h4)")
				},
			],
		}),
	)

	expect(ctx.css("prose-headings:text-black")).toEqual({
		":where(&) :is(h1, h2, h3, h4)": {
			color: "#000",
		},
	})
})

test("matchVariant", async () => {
	const ctx = createContext(
		resolveConfig({
			plugins: [
				({ matchVariant, e }) => {
					matchVariant("tab", value => {
						if (!value) return "& > *"
						return `&.${e(value)} > *`
					})
				},
			],
		}),
	)

	expect(ctx.css("tab-[]:text-black")).toEqual({
		"& > *": {
			color: "#000",
		},
	})

	expect(ctx.css("tab-[test]:text-black")).toEqual({
		"&.test > *": {
			color: "#000",
		},
	})
})
