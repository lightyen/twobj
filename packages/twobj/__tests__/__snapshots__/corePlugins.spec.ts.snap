// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preflight 1`] = `
{
  "*": {
    "borderColor": "#e5e7eb",
    "borderStyle": "solid",
    "borderWidth": "0",
    "boxSizing": "border-box",
  },
  ":-moz-focusring": {
    "outline": "auto",
  },
  ":-moz-ui-invalid": {
    "boxShadow": "none",
  },
  "::-webkit-file-upload-button": {
    "WebkitAppearance": "button",
    "font": "inherit",
  },
  "::-webkit-inner-spin-button,::-webkit-outer-spin-button": {
    "height": "auto",
  },
  "::-webkit-search-decoration": {
    "WebkitAppearance": "none",
  },
  "::after": {
    "--tw-content": "''",
    "borderColor": "#e5e7eb",
    "borderStyle": "solid",
    "borderWidth": "0",
    "boxSizing": "border-box",
  },
  "::before": {
    "--tw-content": "''",
    "borderColor": "#e5e7eb",
    "borderStyle": "solid",
    "borderWidth": "0",
    "boxSizing": "border-box",
  },
  ":disabled": {
    "cursor": "default",
  },
  "[hidden]": {
    "display": "none",
  },
  "[type='search']": {
    "WebkitAppearance": "textfield",
    "outlineOffset": "-2px",
  },
  "a": {
    "color": "inherit",
    "textDecoration": "inherit",
  },
  "abbr:where([title])": {
    "textDecoration": "underline dotted",
  },
  "b,strong": {
    "fontWeight": "bolder",
  },
  "blockquote,dl,dd,h1,h2,h3,h4,h5,h6,hr,figure,p,pre": {
    "margin": "0",
  },
  "body": {
    "lineHeight": "inherit",
    "margin": "0",
  },
  "button,[role="button"]": {
    "cursor": "pointer",
  },
  "button,input,optgroup,select,textarea": {
    "color": "inherit",
    "fontFamily": "inherit",
    "fontFeatureSettings": "inherit",
    "fontSize": "100%",
    "fontVariationSettings": "inherit",
    "fontWeight": "inherit",
    "letterSpacing": "inherit",
    "lineHeight": "inherit",
    "margin": "0",
    "padding": "0",
  },
  "button,input:where([type='button']),input:where([type='reset']),input:where([type='submit'])": {
    "WebkitAppearance": "button",
    "backgroundColor": "transparent",
    "backgroundImage": "none",
  },
  "button,select": {
    "textTransform": "none",
  },
  "code,kbd,samp,pre": {
    "fontFamily": "ui-monospace, SFMono-Regular, Menlo, Monaco, Consolas, "Liberation Mono", "Courier New", monospace",
    "fontFeatureSettings": "normal",
    "fontSize": "1em",
    "fontVariationSettings": "normal",
  },
  "dialog": {
    "padding": "0",
  },
  "fieldset": {
    "margin": "0",
    "padding": "0",
  },
  "h1,h2,h3,h4,h5,h6": {
    "fontSize": "inherit",
    "fontWeight": "inherit",
  },
  "hr": {
    "borderTopWidth": "1px",
    "color": "inherit",
    "height": "0",
  },
  "html,:host": {
    "MozTabSize": "4",
    "WebkitTapHighlightColor": "transparent",
    "WebkitTextSizeAdjust": "100%",
    "fontFamily": "ui-sans-serif, system-ui, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji"",
    "fontFeatureSettings": "normal",
    "fontVariationSettings": "normal",
    "lineHeight": "1.5",
    "tabSize": "4",
  },
  "img,svg,video,canvas,audio,iframe,embed,object": {
    "display": "block",
    "verticalAlign": "middle",
  },
  "img,video": {
    "height": "auto",
    "maxWidth": "100%",
  },
  "input::placeholder,textarea::placeholder": {
    "color": "#9ca3af",
    "opacity": "1",
  },
  "legend": {
    "padding": "0",
  },
  "ol,ul,menu": {
    "listStyle": "none",
    "margin": "0",
    "padding": "0",
  },
  "progress": {
    "verticalAlign": "baseline",
  },
  "small": {
    "fontSize": "80%",
  },
  "sub": {
    "bottom": "-0.25em",
  },
  "sub,sup": {
    "fontSize": "75%",
    "lineHeight": "0",
    "position": "relative",
    "verticalAlign": "baseline",
  },
  "summary": {
    "display": "list-item",
  },
  "sup": {
    "top": "-0.5em",
  },
  "table": {
    "borderCollapse": "collapse",
    "borderColor": "inherit",
    "textIndent": "0",
  },
  "textarea": {
    "resize": "vertical",
  },
}
`;
